File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int n = nums.length;
        int i = 0;
        int j = 0;
        long sum = 0;
        long max = 0;

        HashSet<Integer> set = new HashSet<>();

        while(j < n){
            if(!set.contains(nums[j])){
                set.add(nums[j]);
                sum += nums[j];
                j++;
            }
            else{
                set.remove(nums[i]);
                sum -= nums[i];
                i++;
            }

            if(j - i == k){
                max = Math.max(max, sum);

                set.remove(nums[i]);
                sum -= nums[i];
                i++;
            }
        }

        return max;
    }
}

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        int l = 0;
        int n = nums.size();
        unordered_map<int, int> mp;
        long long curSum = 0;
        long long maxSum = 0;
        for(int r = 0; r < n; r++){
            curSum += nums[r];
            mp[nums[r]]++;
            if((r - l + 1) == k){
                if(mp.size() == k){
                    maxSum = max(maxSum, curSum);
                }
                curSum -= nums[l];
                mp[nums[l]]--;
                if(mp[nums[l]] <= 0){
                    mp.erase(nums[l]);
                }
                l++;
            }
        }  

        return maxSum;
    }
};